Add automated tests for cluster expansion

slaperche-scality opened the pull request.

bert-e commented:
# Hello slaperche-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/982-test-cluster-expansion` into
`development/2.0`
* [w/2.1/feature/982-test-cluster-expansion](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/feature/982-test-cluster-expansion) will be merged into `development/2.1` 
* [w/2.2/feature/982-test-cluster-expansion](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/feature/982-test-cluster-expansion) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf

slaperche-scality commented:
PR not complete yet (doesn't test much of our code yet…), but you can start the review (and it should be complete enough to pass the CI)

gdemonet commented on tests/install/steps/files/node.yaml.tpl:
```diff
@@ -0,0 +1,12 @@
+metadata:
+  name: node1
+  labels:
+    beta.kubernetes.io/os: linux
```
This should not be created by us, `kubelet` takes care of it

slaperche-scality replied:
Ha yes, just copy/pasted it from the Google Doc. I'll remove it!

gdemonet commented on tests/install/steps/test_expansion.py:
```diff
@@ -0,0 +1,61 @@
+    node_ip = get_node_ip(hostname, ssh_config)
+    node_manifest = get_node_manifest(node_ip)
+    k8s.config.load_kube_config(config_file=kubeconfig)
+    k8s_client = k8s.client.CoreV1Api()
```
Minor: we could share such a client between tests of a given session:

```python
@pytest.fixture(scope='session')
def k8s_client(kubeconfig):
    k8s.config.load_kube_config(config_file=kubeconfig)
    return k8s.client.CoreV1Api()
```

slaperche-scality replied:
Clearly! I'll do that

gdemonet commented on tests/install/steps/test_expansion.py:
```diff
@@ -0,0 +1,61 @@
+# }}}
+# Helpers {{{
+
+def get_node_ip(hostname, ssh_config):
```
Minor: could be interesting to mention in the docstring why we needed to use this convoluted approach to retrieve the node IP address

slaperche-scality replied:
Because there is no easy way (that we can find at least) to do this basic task? Or do you have another reason in mind?

gdemonet replied:
We want to avoid running a dumb `ip address show eth1 | awk ...` on the node when we already have that information in the SSH config file. Of course, you could also add a mention that `testinfra` does not provide a clear interface to read from this SSH config file, which is why we used this strange approach. That's what I wanted to see in a comment/docstring.

gdemonet commented on tests/install/steps/test_expansion.py:
```diff
@@ -0,0 +1,61 @@
+# Then {{{
+
+@then(parsers.parse('node "{hostname}" is registered in Kubernetes'))
+def check_node_is_registered(host, request, kubeconfig, hostname):
```
You don't need `host` or `request` here:

```suggestion
def check_node_is_registered(kubeconfig, hostname):
```

slaperche-scality replied:
Indeed.

I thought `host` was mandarory because I saw this in another file:

```
# Scenarios
@scenario('../features/dns_resolution.feature', 'check DNS')
def test_dns(host):
    pass
```

But I guess I can remove it from the `@scenario` in my file as well, no?

gdemonet commented on tests/install/steps/test_expansion.py:
```diff
@@ -0,0 +1,61 @@
+# When {{{
+
+@when(parsers.parse('we declare a new node on host "{hostname}"'))
+def declare_node(host, request, kubeconfig, hostname):
```
You don't need `host` here:

```suggestion
def declare_node(request, kubeconfig, hostname):
```

slaperche-scality replied:
Done.

gdemonet commented on tests/install/steps/test_expansion.py:
```diff
@@ -0,0 +1,61 @@
+
+# Scenarios
+@scenario('../features/expansion.feature', 'Add one node to the cluster')
+def test_cluster_expansion(host):
```
Not sure if you need to specify `host` here, since it seems the `given`/`when`/`then` can ask for "extra" fixtures (e.g. `kubeconfig`).

slaperche-scality replied:
Done

slaperche-scality replied:
Hum, after a quick test it seems required:

```
    @scenario('../features/expansion.feature', 'Add one node to the cluster')
>   def test_cluster_expansion():

tests/install/steps/test_expansion.py:11: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.tox/tests/lib/python3.7/site-packages/pytest_bdd/scenario.py:195: in _execute_scenario
    _execute_step_function(request, scenario, step, step_func)
.tox/tests/lib/python3.7/site-packages/pytest_bdd/scenario.py:136: in _execute_step_function
    step_func(**kwargs)
.tox/tests/lib/python3.7/site-packages/pytest_bdd/steps.py:164: in step_func
    result = request.getfixturevalue(func.__name__)
.tox/tests/lib/python3.7/site-packages/_pytest/fixtures.py:478: in getfixturevalue
    return self._get_active_fixturedef(argname).cached_result[0]
.tox/tests/lib/python3.7/site-packages/_pytest/fixtures.py:501: in _get_active_fixturedef
    self._compute_fixture_value(fixturedef)
.tox/tests/lib/python3.7/site-packages/_pytest/fixtures.py:586: in _compute_fixture_value
    fixturedef.execute(request=subrequest)
.tox/tests/lib/python3.7/site-packages/_pytest/fixtures.py:862: in execute
    fixturedef = request._get_active_fixturedef(argname)
.tox/tests/lib/python3.7/site-packages/_pytest/fixtures.py:501: in _get_active_fixturedef
    self._compute_fixture_value(fixturedef)
.tox/tests/lib/python3.7/site-packages/_pytest/fixtures.py:586: in _compute_fixture_value
    fixturedef.execute(request=subrequest)
.tox/tests/lib/python3.7/site-packages/_pytest/fixtures.py:862: in execute
    fixturedef = request._get_active_fixturedef(argname)
.tox/tests/lib/python3.7/site-packages/_pytest/fixtures.py:501: in _get_active_fixturedef
    self._compute_fixture_value(fixturedef)
.tox/tests/lib/python3.7/site-packages/_pytest/fixtures.py:586: in _compute_fixture_value
    fixturedef.execute(request=subrequest)
.tox/tests/lib/python3.7/site-packages/_pytest/fixtures.py:881: in execute
    return hook.pytest_fixture_setup(fixturedef=self, request=request)
.tox/tests/lib/python3.7/site-packages/pluggy/hooks.py:289: in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
.tox/tests/lib/python3.7/site-packages/pluggy/manager.py:68: in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
.tox/tests/lib/python3.7/site-packages/pluggy/manager.py:62: in <lambda>
    firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
.tox/tests/lib/python3.7/site-packages/_pytest/fixtures.py:923: in pytest_fixture_setup
    result = call_fixture_func(fixturefunc, request, kwargs)
.tox/tests/lib/python3.7/site-packages/_pytest/fixtures.py:782: in call_fixture_func
    res = fixturefunc(**kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

request = <SubRequest '_testinfra_host' for <Function test_cluster_expansion>>

    @pytest.fixture(scope="module")
    def _testinfra_host(request):
>       return request.param
E       AttributeError: 'SubRequest' object has no attribute 'param'
```

gdemonet commented on tests/install/steps/test_expansion.py:
```diff
@@ -0,0 +1,61 @@
+
+# When {{{
+
+@when(parsers.parse('we declare a new node on host "{hostname}"'))
```
We may want to include some way to specify the node roles here, using `node-role.kubernetes.io/<some_role>` annotations in the `node_manifest`.

slaperche-scality replied:
Can be introduced when the need occurs.

gdemonet reviewed (commented):
Few minor comments, seems like we're on good tracks to have the expansion tested in the CI!

Small question: do you plan on extending the `Add one node to the cluster` scenario:

```
Scenario: Add one node to the cluster
    Given the Kubernetes API is available
    When we declare a new node on host "node1"
    And we deploy the node named "node1"
    Then node "node1" is registered in Kubernetes
    And node "node1" is running
```

Or would you split it into two scenarii? I'd advocate the latter, although I have no strong opinion.

slaperche-scality commented:
> Few minor comments, seems like we're on good tracks to have the expansion tested in the CI!
> 
> Small question: do you plan on extending the `Add one node to the cluster` scenario:
> 
> ```
> Scenario: Add one node to the cluster
>     Given the Kubernetes API is available
>     When we declare a new node on host "node1"
>     And we deploy the node named "node1"
>     Then node "node1" is registered in Kubernetes
>     And node "node1" is running
> ```
> 
> Or would you split it into two scenarii? I'd advocate the latter, although I have no strong opinion.

I'll go with the former because I want scenario to be self contained: having scenario B requiring the execution of scenario A first is a pain (been there done that).

bert-e commented:
# History mismatch



Merge commit #8a7c37dad78449aad8cfacaea408d928a111e70f on the integration branch
`w/2.1/feature/982-test-cluster-expansion` is merging a branch which is neither the current
branch `feature/982-test-cluster-expansion` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/982-test-cluster-expansion` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

slaperche-scality committed:
tests: add steps to deploy a new node

`strict_gherkin=False` is required to be able to intersperse When and
Then steps (which is allowed by the Gherkin specs anyway)

Refs: #982
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
tests: add a step to check master role

For now, we consider that a node is a master node if kube-scheduler
kube-controller-manager and kube-apiserver are up and running on this
node.

Refs: #982
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
tests: add a step to check etcd role

We consider that a node has the etcd role if:
- etcd pod is up and running
- the node is part of the etcd cluster

Refs: #982
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality commented:
Ok, we currently can't run test on multiple node in the CI (see #1038) so I'll disable the test for now (we can still run them locally, it's better than nothing) so that we can merge this PR.

bert-e commented:
# History mismatch



Merge commit #bcde1d73a58bb5f0fd5d290f83e9a3b8ffe798a5 on the integration branch
`w/2.1/feature/982-test-cluster-expansion` is merging a branch which is neither the current
branch `feature/982-test-cluster-expansion` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/982-test-cluster-expansion` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented on tests/conftest.py:
```diff
@@ -10,3 +15,68 @@ def pytest_addoption(parser):
+
+# Fixtures {{{
+
+@pytest.fixture
```
I know you just moved the code, but this one can have a `scope="module"` as well I believe :)

slaperche-scality replied:
Done

gdemonet commented on tests/install/steps/test_expansion.py:
```diff
@@ -0,0 +1,146 @@
+    assert len(nodes) == 1, '`{}` not registered in Kubernetes'.format(hostname)
+
+
+@then(parsers.parse('node "{node_name}" has the role master'))
```
I'm not sure if this `Then` means what it does... To me, "having the master role" means having the appropriate `node-role.kubernetes.io/master` label.
This `Then` is more about checking that deploying a node with the master role effectively schedules `k-s`, `k-c-m` and `k-a` on it.

Same goes for the notion of `etcd` role.

slaperche-scality replied:
Done

gdemonet commented on tests/install/steps/test_expansion.py:
```diff
@@ -0,0 +1,146 @@
+
+
+def get_node_manifest(metalk8s_version, node_ip):
+    """Return the YAML to declare a node with the specified IP."""
```
Let's add a mention somewhere (either here or simply in the template name and the relevant function/`Given` names) that this is generating a **control-plane `Node`** (i.e. with the `master` and `etcd` roles). Future work could either make these roles customizable in the template (or in this function), or have different templates.

slaperche-scality replied:
Done

gdemonet reviewed (changes requested):
Most of it looks good, I'd just like the scenario to be reworked a bit to look like this:

```
Given we declare a new Node on host "toto" with roles "master,etcd"
Then the Node "toto" is registered
When we deploy the node "toto"
Then we have 1 running Pod labeled "component=kube-controller-manager" on node "toto"
And we have 1 running Pod labeled "component=kube-apiserver" on node "toto"
And we have 1 running Pod labeled "component=kube-scheduler" on node "toto"
And we have 1 running Pod labeled "component=etcd" on node "toto"
```

You could look at the `tests/post/features/pods_alive.feature` for an example of this kind of `Then` usage (check running pods), along with the `tests/kube_utils.py:get_pods` utility function for retrieving pods from a set of labels/namespace/status.

slaperche-scality committed:
tests: make `get_node_manifest` more generic

Now we can specify the template to load (this will make easier to test
the expansion with other kind of nodes).

Refs: #982
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality commented on tests/install/steps/test_expansion.py:
```diff
@@ -0,0 +1,135 @@
+    assert len(nodes) <= 1, 'too many nodes with hostname `{}`'.format(hostname)
+    assert len(nodes) == 1, '`{}` not registered in Kubernetes'.format(hostname)
+
+@then(parsers.parse(
```
That's a copy paste from `post` features and I'm not happy with that.

@gdemonet any way to reuse steps between test suites?

gdemonet replied:
Yep, just define them in a `conftest.py` to make sure they are available. `given`/`when`/`then` are actually fixtures, IIUC.

See https://pytest-bdd.readthedocs.io/en/stable/#reusing-steps

slaperche-scality replied:
Yeah of course, I even already did it for `@given("the Kubernetes API is available")` >_<

gdemonet reviewed (dismissed):
LGTM - tested on local deployment, works fine except for a few details (mine not updated flakiness, kubectl required for install, conflicts if node object already exists). I'll create follow-up tickets for those, should be good to merge as-is.

_Edit_: follow-up tickets:

- Mine flakiness: already tracked in #1044 
- `kubectl` dependency: tracked in #850 (and/or #622)
- `Node` object already exists: tracked in #1056

slaperche-scality commented:
@bert-e approve

bert-e commented:
# History mismatch



Merge commit #bcde1d73a58bb5f0fd5d290f83e9a3b8ffe798a5 on the integration branch
`w/2.1/feature/982-test-cluster-expansion` is merging a branch which is neither the current
branch `feature/982-test-cluster-expansion` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/982-test-cluster-expansion` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

slaperche-scality commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/982-test-cluster-expansion` into
`development/2.0`
* [w/2.1/feature/982-test-cluster-expansion](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/feature/982-test-cluster-expansion) will be merged into `development/2.1` 
* [w/2.2/feature/982-test-cluster-expansion](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/feature/982-test-cluster-expansion) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf










*The following options are set:* **approve**

gdemonet commented on tests/install/features/expansion.feature:
```diff
@@ -0,0 +1,13 @@
+    Scenario: Add one node to the cluster
+        Given the Kubernetes API is available
+        When we declare a new "control-plane" node on host "node1"
+        Then node "node1" is registered in Kubernetes
```
TODO: check status of Node object (before deployment -> `NotReady`, after -> `Ready`)

slaperche-scality replied:
Indeed, would make sense

slaperche-scality replied:
Done

gdemonet commented on tests/install/features/expansion.feature:
```diff
@@ -0,0 +1,13 @@
+        Then we have 1 running pod labeled 'component=kube-controller-manager' on node 'bootstrap'
+        And we have 1 running pod labeled 'component=kube-scheduler' on node 'bootstrap'
+        And we have 1 running pod labeled 'component=kube-apiserver' on node 'bootstrap'
+        And we have 1 running pod labeled 'component=etcd' on node 'bootstrap'
```
Do we need components like "kube-proxy" and some CNI plugin (see #1089)?

slaperche-scality replied:
Done

gdemonet reviewed (commented):
Few notes to remember about before merging

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @jbertran

* @monpote

* @alxf

* @gdemonet

* @lucieleonard

* @zempashi

* @carlito-scality

* @alexandre-allard-scality

* @nootal

* @nicolast

* @thomasdanan

* @tcarmet

* @slaperche-scality

* @teddyandrieux

* @sayf-eddine-scality

* @ramanduh










*The following options are set:* **approve**

slaperche-scality commented:
@bert-e help

bert-e commented:
# Help page



The following options and commands are available at this time.

## Options ##


name   | description  | privileged   | authored
------ | ------------ | ------------ |----------
:arrow_right: **after_pull_request** | Wait for the given pull request id to be merged before continuing with the current one. |  | 
:arrow_right: **bypass_jira_check** | Bypass the Jira issue check |  :star:  | 
:arrow_right: **wait** | Instruct Bert-E not to run until further notice. |  | 
:arrow_right: **bypass_leader_approval** | Bypass the pull request leaders' approval |  :star:  | 
:arrow_right: **bypass_build_status** | Bypass the build and test status |  :star:  | 
:arrow_right: **approve** | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
:arrow_right: **no_octopus** | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
:arrow_right: **bypass_author_approval** | Bypass the pull request author's approval |  :star:  | 
:arrow_right: **bypass_incompatible_branch** | Bypass the check on the source branch prefix |  :star:  | 
:arrow_right: **bypass_commit_size** | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
:arrow_right: **unanimity** | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
:arrow_right: **create_pull_requests** | Allow the creation of integration pull requests. |  | 
:arrow_right: **bypass_peer_approval** | Bypass the pull request peers' approval |  :star:  | 



## Commands ##


name   | description  | privileged
------ | ------------ | ------------
:arrow_right: **force_reset** | Delete integration branches & pull requests, and restart merge process from the beginning. | 
:arrow_right: **status** | Print Bert-E's current status in the pull request ```TBA``` | 
:arrow_right: **help** | Print Bert-E's manual in the pull request. | 
:arrow_right: **reset** | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
:arrow_right: **retry** | Re-start a fresh build ```TBA``` | 
:arrow_right: **build** | Re-start a fresh build ```TBA``` | 
:arrow_right: **clear** | Remove all comments from Bert-E from the history ```TBA``` | 






*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @jbertran

* @monpote

* @alxf

* @gdemonet

* @lucieleonard

* @zempashi

* @carlito-scality

* @alexandre-allard-scality

* @nootal

* @nicolast

* @thomasdanan

* @tcarmet

* @slaperche-scality

* @teddyandrieux

* @sayf-eddine-scality

* @ramanduh










*The following options are set:* **approve**

gdemonet commented on scripts/fix_vagrant_ssh_config.py:
```diff
@@ -0,0 +1,58 @@
+    We return the first IP on `eth1`.
+    """
+    stdout = subprocess.check_output([
+        'vagrant', 'ssh', hostname,
```
TODO: since we retrieve the IP ourselves, let's do it in a fixture instead

_Rationale: IPs in SSH config are not necessarily (even in the Terraform generated ones) the control-plane IP_

slaperche-scality replied:
No longer relevant with your new code.

bert-e commented:
# Branches have diverged



This pull request's source branch `feature/982-test-cluster-expansion` has diverged from
`development/2.0` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.0` into `feature/982-test-cluster-expansion`
* Rebase `feature/982-test-cluster-expansion` onto `origin/development/2.0`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

slaperche-scality committed:
ci: don't run multinodes tests on single node env

Refs: #982
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
ci: dump the SSH config of every node

This will be useful when we want to get the IP from the hostname in our
test suite.

Refs: #982
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
tests: move Kube API check to an upper level

These steps will be needed by the tests under `install` as well.

Refs: #982
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
tests: test the first step of the expansion

Add steps to test the first step of the cluster expansion: declaring the
new node in the Kubernetes API.

Next commit will test the deployment, using the Salt orchestration.

Refs: #982
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
tests: display the error when commands fail

This will ease troubleshooting.

Refs: #982
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
tests: refactor pod listing

In `post` tests we were checking running pods, but I also re-implemented
it for `install` tests…

Code duplication being bad, let's refactor this.
The `post` tests were calling `kubectl`, whereas my tests where using
the API directly.

So, I rewrote the existing helper (using API) and also added the ability
to filter by node (which allows to use `==` instead of `>=`).

Note that since we are now filtering by node name, we need to resolve
the node name from the SSH config to a real hostname known by k8s.

Refs: #982
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

bert-e commented:
# History mismatch



Merge commit #be6618ed18145fcccc0889eb0e7e9e89988430b4 on the integration branch
`w/2.1/feature/982-test-cluster-expansion` is merging a branch which is neither the current
branch `feature/982-test-cluster-expansion` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/982-test-cluster-expansion` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

slaperche-scality commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/982-test-cluster-expansion` into
`development/2.0`
* [w/2.1/feature/982-test-cluster-expansion](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/feature/982-test-cluster-expansion) will be merged into `development/2.1` 
* [w/2.2/feature/982-test-cluster-expansion](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/feature/982-test-cluster-expansion) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @alexandre-allard-scality

* @gdemonet

* @zempashi

* @sayf-eddine-scality

* @slaperche-scality

* @alxf

* @monpote

* @tcarmet

* @carlito-scality

* @ramanduh

* @cmonfort

* @thomasdanan

* @lucieleonard

* @nicolast

* @teddyandrieux

* @nootal

bert-e commented:
# History mismatch



Merge commit #1c5736e1119017e9e585be96c4e260f9dcc3b2c9 on the integration branch
`w/2.1/feature/982-test-cluster-expansion` is merging a branch which is neither the current
branch `feature/982-test-cluster-expansion` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/982-test-cluster-expansion` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

slaperche-scality commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/982-test-cluster-expansion` into
`development/2.0`
* [w/2.1/feature/982-test-cluster-expansion](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/feature/982-test-cluster-expansion) will be merged into `development/2.1` 
* [w/2.2/feature/982-test-cluster-expansion](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/feature/982-test-cluster-expansion) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @alexandre-allard-scality

* @gdemonet

* @zempashi

* @sayf-eddine-scality

* @slaperche-scality

* @alxf

* @monpote

* @tcarmet

* @carlito-scality

* @ramanduh

* @cmonfort

* @thomasdanan

* @lucieleonard

* @nicolast

* @teddyandrieux

* @nootal

slaperche-scality committed:
tests: add check on the node status

Closes: #982
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

bert-e commented:
# History mismatch



Merge commit #29b533fb4310dd7bb08c24f81fdb5e5b8f384358 on the integration branch
`w/2.1/feature/982-test-cluster-expansion` is merging a branch which is neither the current
branch `feature/982-test-cluster-expansion` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/982-test-cluster-expansion` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

slaperche-scality commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/982-test-cluster-expansion` into
`development/2.0`
* [w/2.1/feature/982-test-cluster-expansion](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/feature/982-test-cluster-expansion) will be merged into `development/2.1` 
* [w/2.2/feature/982-test-cluster-expansion](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/feature/982-test-cluster-expansion) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @alexandre-allard-scality

* @gdemonet

* @zempashi

* @sayf-eddine-scality

* @slaperche-scality

* @alxf

* @monpote

* @tcarmet

* @carlito-scality

* @ramanduh

* @cmonfort

* @thomasdanan

* @lucieleonard

* @nicolast

* @teddyandrieux

* @nootal

bert-e commented:
# History mismatch



Merge commit #a82225a4394f091ca339efeb7dab5c5fd54cc48d on the integration branch
`w/2.1/feature/982-test-cluster-expansion` is merging a branch which is neither the current
branch `feature/982-test-cluster-expansion` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/982-test-cluster-expansion` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

slaperche-scality committed:
ci: deploy the SSH key pair for expansion

We deploy the public key on node1 and the private key on the bootstrap
node, so that the key pair can be used for cluster expansion.
Note that since we can't connect as root, we have to use two steps to
copy the private key in the right location (/etc/metalk8s/pki/).

This is a temporary workaround until we do it TheRightWay™ by using
Terraform.

Refs: #982, #1222
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
tests: don't resolve hostname in single-node tests

In the single-node test environment, we only have the bootstrap node and
we force its hostname to "bootstrap" so there is no need to resolve it.
Moreover, trying to do so ends up with the tests failures (my guess it
that when giving an `ssh_config` set to `None` testinfra/paramiko still
try to use SSH, but fallback on a default configuration that won't work.

Given the fact that `ssh_config` is not set in the single-node test
environment, we use its value to know if we need to resolve the hostname
or not.

Refs: #982
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
tests: add hostname resolution where needed

In several place we were using hostname from the SSH config, whereas we
should use the real one (since it's the one known by k8s).

Refs: #982
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
tests: add infra label

Otherwise Calico is not properly configured and `test_dns` fails when
scheduling the busybox on `node1`.

Refs: #982
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
ci: move control and workload plane into tenant network

Since we're having issue with the `Port Security` feature of OpenStack
and that's it kind of harder than expected to workaround, we will
temporarily disable the dedicated network for the control and workload
plane and enable IPIP.

Refs: #982

bert-e commented:
# Branches have diverged



This pull request's source branch `feature/982-test-cluster-expansion` has diverged from
`development/2.0` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.0` into `feature/982-test-cluster-expansion`
* Rebase `feature/982-test-cluster-expansion` onto `origin/development/2.0`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

bert-e commented:
# History mismatch



Merge commit #a82225a4394f091ca339efeb7dab5c5fd54cc48d on the integration branch
`w/2.1/feature/982-test-cluster-expansion` is merging a branch which is neither the current
branch `feature/982-test-cluster-expansion` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/982-test-cluster-expansion` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

slaperche-scality commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/982-test-cluster-expansion` into
`development/2.0`
* [w/2.1/feature/982-test-cluster-expansion](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/feature/982-test-cluster-expansion) will be merged into `development/2.1` 
* [w/2.2/feature/982-test-cluster-expansion](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/feature/982-test-cluster-expansion) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet

bert-e commented:
# History mismatch



Merge commit #ee73c7c6da92684b58be8fdca52c36b51a93a244 on the integration branch
`w/2.1/feature/982-test-cluster-expansion` is merging a branch which is neither the current
branch `feature/982-test-cluster-expansion` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/982-test-cluster-expansion` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

bert-e commented:
# History mismatch



Merge commit #222f0b4c6a3384b45022d534f20a1e77f2e36ee5 on the integration branch
`w/2.1/feature/982-test-cluster-expansion` is merging a branch which is neither the current
branch `feature/982-test-cluster-expansion` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/982-test-cluster-expansion` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

bert-e commented:
# History mismatch



Merge commit #620feca9b130df56b6c1da04bea6739e68792b48 on the integration branch
`w/2.1/feature/982-test-cluster-expansion` is merging a branch which is neither the current
branch `feature/982-test-cluster-expansion` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/982-test-cluster-expansion` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

NicolasT commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/982-test-cluster-expansion` into
`development/2.0`
* [w/2.1/feature/982-test-cluster-expansion](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/feature/982-test-cluster-expansion) will be merged into `development/2.1` 
* [w/2.2/feature/982-test-cluster-expansion](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/feature/982-test-cluster-expansion) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet

NicolasT commented:
Rebasing to get the CI fixes...

bert-e commented:
# History mismatch



Merge commit #5f83c976cb2f0bd2634d210918bc0a3e2c3e8e62 on the integration branch
`w/2.1/feature/982-test-cluster-expansion` is merging a branch which is neither the current
branch `feature/982-test-cluster-expansion` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/982-test-cluster-expansion` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

NicolasT commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/982-test-cluster-expansion` into
`development/2.0`
* [w/2.1/feature/982-test-cluster-expansion](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/feature/982-test-cluster-expansion) will be merged into `development/2.1` 
* [w/2.2/feature/982-test-cluster-expansion](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/feature/982-test-cluster-expansion) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet

NicolasT commented:
@bert-e help

bert-e commented:
# Help page



The following options and commands are available at this time.

## Options ##


name   | description  | privileged   | authored
------ | ------------ | ------------ |----------
:arrow_right: **bypass_build_status** | Bypass the build and test status |  :star:  | 
:arrow_right: **bypass_author_approval** | Bypass the pull request author's approval |  :star:  | 
:arrow_right: **bypass_peer_approval** | Bypass the pull request peers' approval |  :star:  | 
:arrow_right: **approve** | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
:arrow_right: **wait** | Instruct Bert-E not to run until further notice. |  | 
:arrow_right: **bypass_commit_size** | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
:arrow_right: **bypass_jira_check** | Bypass the Jira issue check |  :star:  | 
:arrow_right: **create_pull_requests** | Allow the creation of integration pull requests. |  | 
:arrow_right: **unanimity** | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
:arrow_right: **after_pull_request** | Wait for the given pull request id to be merged before continuing with the current one. |  | 
:arrow_right: **no_octopus** | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
:arrow_right: **bypass_leader_approval** | Bypass the pull request leaders' approval |  :star:  | 
:arrow_right: **bypass_incompatible_branch** | Bypass the check on the source branch prefix |  :star:  | 



## Commands ##


name   | description  | privileged
------ | ------------ | ------------
:arrow_right: **status** | Print Bert-E's current status in the pull request ```TBA``` | 
:arrow_right: **clear** | Remove all comments from Bert-E from the history ```TBA``` | 
:arrow_right: **reset** | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
:arrow_right: **help** | Print Bert-E's manual in the pull request. | 
:arrow_right: **build** | Re-start a fresh build ```TBA``` | 
:arrow_right: **retry** | Re-start a fresh build ```TBA``` | 
:arrow_right: **force_reset** | Delete integration branches & pull requests, and restart merge process from the beginning. |

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet

NicolasT commented:
@bert-e bypass_author_approval

NicolasT commented:
@bert-e status

bert-e commented:
# Status





*Status report is not available.*






*The following options are set:* **bypass_author_approval**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/21186c539b72f74142d99c07d7b32fba66a5c68f) did not succeed in branch w/2.2/feature/982-test-cluster-expansion.




*The following options are set:* **bypass_author_approval**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **bypass_author_approval**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



Please check the status of the associated issue None.

Goodbye slaperche-scality.
